specification.md

```diff
@@ -773,6 +773,59 @@ Path? getStderrPath()
 
 Set/get the paths to the standard stream files.
 
+<a name="jobspec-setprelaunch"></a>
+<a name="jobspec-setpreexec"></a>
+<a name="jobspec-setpostexec"></a>
+<a name="jobspec-setpostlaunch"></a>
+```java
+void setPreLaunch(Path script)
+Path? getPreLaunch()
+void setPostLaunch(Path script)
+Path? getPostLaunch()
+```
+
+Set/get the paths to the pre/post launch scripts.
+
+The `PreLaunch` script is sourced by the main process of the job before it uses
+the parallel launcher to execute the `executable`.  This script is intended for
+setting up the environment via external systems like `lmod` and `virtualenv` as
+well as for performing actions that should only be done once at the start of a
+job (e.g., creating a results directory).
+
```

Comment:
Thanks, this reads good to me.

One problem we repeatedly faced with our users is the assumption that `pre_launch` scripts run concurrently on the compute nodes.  Well, they don't - at best they run concurrently on the batch node, at worst they run serial on the login node...  We might want to document this - maybe not here in the spec though, but rather in the implementation docs?

Reply:
What are your thoughts on adding to the user note below something like the following:

```
Since the `PreLaunch` script is sourced before the parallel launch, it is only sourced by a single process.  This process may be running on the login node or on a batch node, depending on the system and J/PSI implementation.
```

Reply:
I just pushed up a fixup commit with the suggested change above (plus an extra sentence or two).  Let me know what you think.  Happy to modify or back it out based on feedback.